### ISSUE TITLE
[Temporary fix] for Expected quality not found! error

### ISSUE THREAD

**Issue by IronSuit** [nishad/udemy-dl-windows#143]
Hello everyone,

This fix is made possible by @smeggingsmegger to solve the Expected quality not found issue.
It took me nearly 5 hours to figure this method, this tutorial for people like me who don't know anything in regards to Python or any coding skills.

1. download python 2.7 from [https://www.python.org/downloads/ ](url)
2. Install python 2.7 but don't change the install directory and leave it as it is which is "C:\Python27\"
3. download and install "setuptools-0.6c11.win32-py2.7.exe" from [https://pypi.python.org/pypi/setuptools/0.6c11](url)
4. Open cmd and type the following `pushd C:\Python27\Scripts`
5. type the following command to install pip `easy_install pip`
6. download the git from [https://git-scm.com/download/win](url)
7. close and reopen cmd and type the following `pushd C:\Python27\Scripts`
8. Type `pip install -U git+https://github.com/smeggingsmegger/udemy-dl-windows.git@master` to download the fix
9. go to "C:\Python27" and create a new batch and name it `udemy-dl.bat` and write inside it 
```
@echo off
for /f %%p in ('where python.exe') do set pythonpath=%%p
set "_python=%pythonpath%"
for %%a in ("%_python%") do set "python_scripts=%%~dpa\scripts"
python.exe %python_scripts%\udemy-dl %*
```

step 9 is credited to @mrswadge

10. in cmd type the following `pushd C:\Python27`
11. now everything should be working as usual, so to download just type as normal `udemy-dl COURSELINK`

Please note that the above method worked for me and i cannot guarantee that it will work with everyone, but hey it's better than nothing.
Also this method is for Microsoft windows.

**Comment by aolko** [comment]
please push the fixed version
step 8 fails dramatically

**Comment by hariwyzmindz** [comment]
yes, facing the same issue

**Comment by hariwyzmindz** [comment]
C:\Python27\Scripts>pip install -U git+https://github.com/smeggingsmegger/udemy-dl-windows.git@master
Collecting git+https://github.com/smeggingsmegger/udemy-dl-windows.git@master
  Cloning https://github.com/smeggingsmegger/udemy-dl-windows.git (to master) to c:\users\hari\appdata\local\temp\pip-i0b9_z-build
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    IOError: [Errno 2] No such file or directory: 'c:\\users\\hari\\appdata\\local\\temp\\pip-i0b9_z-build\\setup.py'

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in c:\users\hari\appdata\local\temp\pip-i0b9_z-build\

**Comment by IronSuit** [comment]
@aolko i don't even know how to push the fix, i don't think its possible because it downloaded tons of folders for me.

@aolko @hariwyzmindz 
try to type `pip install udemy-dl` in step 8

**Comment by aolko** [comment]
it won't download videos

**Comment by hariwyzmindz** [comment]
yep, doesn't work

**Comment by Eshpelin** [comment]
Getting this on Step 8:

IOError: [Errno 2] No such file or directory: 'c:\users\eshpelin\appdata\local\temp\pip-i0b9_z-build\setup.py'

**Comment by aolko** [comment]
nope, iwould suggest to get it thru pip instead and THEN run the .bat, but it wont work anyway

**Comment by toeknee581** [comment]
I'm new to contributing to FOSS, but I have been able to get this to work.  I'll try to explain and then maybe tomorrow I can learn how to do a pull request.

First, I noticed that he bug has been fixed and committed, but the binary (.exe) has not been updated in the dist.  Someone in another thread was asking how to build it yourself.

1) Download Python 2.7.13 from python.org.  If you use the installer it will set up the path for you
2) Use PIP (which was installed with Python) at a command prompt to install PythonInstaller with command "pip install pyinstaller"
3) Install the other required libraries the same way.  These are listed in requirements.txt.  Commands:
pip install requests==2.9.1
pip install unicode-slugify
pip install colorlog[windows]==2.7.0
pip install future==0.15.2
pip install colorama
4) in the master folder, you will see a file called udemy-dl.spec.  One edit needs to be made to the third line.  Make it hiddenimports=["requests"], (or you will get an error when you run it about the requests module not being found)
5) edit the udemy-dl.py file in the src directory.  around line 19 with the other imports, add "import json"
6) go back to the main folder with the .spec file and use the command "pyinstaller udemy-dl.spec" which will build the executable.  I noticed sometimes that I had to delete the .exe from the dist folder for it to update.
7) Your shiny new exe should now exist in your dist folder and it should work!

Enjoy the learning - I spend tons of time on planes and this project has helped me study in the sky.  Thanks!

CoderTonyB

**Comment by aolko** [comment]
[Working version download link](https://drop.me/oWVOmB)

**Comment by toeknee581** [comment]
Nice aolko - I see you were able to make the changes and initiate a pull request.  It looks like the exe doesn't get updated through the pull request though.  Any idea how we do that?

**Comment by nishad** [comment]
@aolko Providing external links to binary files are not permitted for the safety and security of users of this project. So your previous comment is removed from the thread. Anyway thanks for your efforts.

**Comment by aolko** [comment]
requesting new release then

**Comment by JacquesvanWyk** [comment]
@nishad when will this be fixed? I am not getting above instructions done correctly i dont have much python experience.

**Comment by shyamsukhamit** [comment]
Any updates on this windows fix?

**Comment by shadsluiter** [comment]
Can anyone push the EXE version of this program to the dist folder?

**Comment by aolko** [comment]
my pull requests were not accepted anyway, so...

**Comment by nishad** [comment]
@aolko I guess, still there are a lot of users who got confused with what this project is all about.

This project is a Windows binary distribution of https://pypi.python.org/pypi/udemy-dl. 

Windows standalone binaries are provided to help Windows users to avoid the pain of installing Python and all dependencies. Which also helps novice users from debugging Python related issues. 

If there is a new release for https://pypi.python.org/pypi/udemy-dl, this project will use that to produce new windows binaries (or `udemy-dl.exe`, as you call it).

**Comment by smeggingsmegger** [comment]
@nishad are you considering releasing my patch to udemy-dl? I am getting so many pings about this that I'd like to be done one way or the other. :)

**Comment by nishad** [comment]
@smeggingsmegger Usually it takes 3-4 weeks for fixing any udemy.com website related changes as udemy tends to make multiple changes in a span and takes some time to update their website globally. `udemy-dl` is not a mission critical project to issue hot patches since all purchases are life long and users can download them anytime later.

**Comment by smeggingsmegger** [comment]
@nishad thank you! It's a testament to how popular and beloved your project is that I am getting so many comments about my patch. It might help to post that on the PR / Issue to stem the flow of questions and comments. Thanks again for making this utility!

**Comment by MoCityMM** [comment]
Getting the same error but, also getting SSL certificate errors from 'curl'... don't know if that mean anything to anyone.
'curl: (60) SSL certificate problem: unable to get local issuer certificate'

**Comment by CaptainMalzbier** [comment]
Hey guys, found a solution: 

IN YOUR CMD:
`pip install youtube-dl`

Then change to your download directory 

`youtube-dl URLToYourCourse -u username -p password --all-subs`

and then I've created an index.html and added the following:

`<video id="movie1" controls preload="metadata">
 <source src="yourVideo.mp4"/>
 <track kind="subtitles" label="EN subtitles" src="yourVideoSubtitle.en.vtt" srclang="en" default/>   
        Your browser doesn't support HTML5. Maybe you should upgrade.
</video>`

And it worked with Mozilla Firefox (for me)

**Comment by nishad** [comment]
Fixed with v2.1.3 (https://github.com/nishad/udemy-dl-windows/releases/tag/v2.1.3)
